tests/client/test_client.py

```diff
@@ -153,3 +153,54 @@ def test_elapsed_delay(server):
     with httpx.Client() as client:
         response = client.get(url)
     assert response.elapsed.total_seconds() > 0.0
+
+
+_wsgi_body = b"""
+
+
+<!DOCTYPE html>
+<html>
+  <head>
+    <title>This is not a web page</title>
+  </head>
+  <body>
+    <p>This page deliberately left blank</p>
+  </body>
+</html>
+"""
+
+
+def test_wsgi_app():
+    def application(env, start_response):
+        start_response("200 OK", [("Content-Type", "text/html")])
+        return [_wsgi_body]
+
+    with httpx.Client(app=application) as client:
+        response = client.get("http://example.com/")
+        assert response.status_code == 200
+        assert b"This page deliberately left blank" in response.content
+
+
+def test_wsgi_app_generator():
+    def application(env, start_response):
+        start_response("200 OK", [("Content-Type", "text/html")])
+        for line in _wsgi_body.split(b"\n"):
+            yield line
+
+    with httpx.Client(app=application) as client:
+        response = client.get("http://example.com/")
+        assert response.status_code == 200
+        assert b"This page deliberately left blank" in response.content
+
+
+def test_wsgi_app_generator_empty():
+    def application(env, start_response):
+        body = [b"", b"", b""]
+        start_response("200 OK", [("Content-Type", "text/html")])
+        for line in body:
+            yield line
+
+    with httpx.Client(app=application) as client:
+        response = client.get("http://example.com/")
+        assert response.status_code == 200
+        assert response.content == b""
```

Comment:
FYI, we already have a `tests/test_wsgi.py` file. :-)

It doesn't seem like we need more tests there — we'd just need to refactor `echo_body_with_response_stream` to be an actual generator (instead of calling `start_response` synchronously and *then* returning a generator), which would match eg the `test_wsgi_app_generator` case here:

https://github.com/encode/httpx/blob/fc980e779276e46ce2e2aad10a8ced1d50334306/tests/test_wsgi.py#L36-L50

Reply:
I'll move them.  Though I'm a little surprised the tests aren't in `tests/client/`.  I assumed the test structure would roughly match the code structure.

Reply:
Yeah it's a bit surprising. Personally I'd see them in `tests/dispatch/` since `asgi.py` and `wsgi.py` are in the `_dispatch` sub-package. :-) Happy to review any follow-up PR for this!